### DISCUSSION TITLE
Module 'psycopg2' Not Found

### DISCUSSION THREAD

**Discussion by seejayyy** [frePPLe/frepple#616] [Q&A] [answered]
Hello,

I have recently installed frePPLe software and got it running on Ubuntu 22.04.3.

I followed the Demand Forecasting demo video on YouTube and I populated the Items, Locations, User, Sales Order, Forecast tables using the demo values. However, when I tried to go Admin > Execute, it returns me Server Error (500). 

Upon checking the log file, I saw this error 

`Traceback (most recent call last):
[Thu Nov 09 04:26:45.689953 2023] [wsgi:error] [pid 40626:tid 139688744765312] Traceback (most recent call last):
[Thu Nov 09 04:26:45.689963 2023] [wsgi:error] [pid 40622:tid 139688744765312] Traceback (most recent call last):
[Thu Nov 09 04:26:45.689974 2023] [wsgi:error] [pid 40623:tid 139688744765312]   File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/../django/db/backends/postgresql/base.py", line 25, in <>
[Thu Nov 09 04:26:45.689974 2023] [wsgi:error] [pid 40625:tid 139688744765312] Traceback (most recent call last):
[Thu Nov 09 04:26:45.689978 2023] [wsgi:error] [pid 40623:tid 139688744765312]     import psycopg2 as Database
[Thu Nov 09 04:26:45.689978 2023] [wsgi:error] [pid 40626:tid 139688744765312]   File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/../django/db/backends/postgresql/base.py", line 25, in <>
[Thu Nov 09 04:26:45.689984 2023] [wsgi:error] [pid 40622:tid 139688744765312]   File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/../django/db/backends/postgresql/base.py", line 25, in <>
[Thu Nov 09 04:26:45.689989 2023] [wsgi:error] [pid 40623:tid 139688744765312]   File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/../psycopg2/__init__.py", line 51, in <module>
[Thu Nov 09 04:26:45.689997 2023] [wsgi:error] [pid 40626:tid 139688744765312]     import psycopg2 as Database
[Thu Nov 09 04:26:45.690000 2023] [wsgi:error] [pid 40625:tid 139688744765312]   File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/../django/db/backends/postgresql/base.py", line 25, in <>
[Thu Nov 09 04:26:45.690001 2023] [wsgi:error] [pid 40622:tid 139688744765312]     import psycopg2 as Database
[Thu Nov 09 04:26:45.690003 2023] [wsgi:error] [pid 40623:tid 139688744765312]     from psycopg2._psycopg import (                     # noqa
[Thu Nov 09 04:26:45.690007 2023] [wsgi:error] [pid 40626:tid 139688744765312]   File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/../psycopg2/__init__.py", line 51, in <module>
[Thu Nov 09 04:26:45.690009 2023] [wsgi:error] [pid 40625:tid 139688744765312]     import psycopg2 as Database
[Thu Nov 09 04:26:45.690013 2023] [wsgi:error] [pid 40622:tid 139688744765312]   File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/../psycopg2/__init__.py", line 51, in <module>
[Thu Nov 09 04:26:45.690015 2023] [wsgi:error] [pid 40626:tid 139688744765312]     from psycopg2._psycopg import (                     # noqa
[Thu Nov 09 04:26:45.690020 2023] [wsgi:error] [pid 40623:tid 139688744765312] ModuleNotFoundError: No module named 'psycopg2._psycopg'
[Thu Nov 09 04:26:45.690020 2023] [wsgi:error] [pid 40622:tid 139688744765312]     from psycopg2._psycopg import (                     # noqa
[Thu Nov 09 04:26:45.690021 2023] [wsgi:error] [pid 40625:tid 139688744765312]   File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/../psycopg2/__init__.py", line 51, in <module>
[Thu Nov 09 04:26:45.690026 2023] [wsgi:error] [pid 40623:tid 139688744765312]
[Thu Nov 09 04:26:45.690029 2023] [wsgi:error] [pid 40626:tid 139688744765312] ModuleNotFoundError: No module named 'psycopg2._psycopg'
[Thu Nov 09 04:26:45.690029 2023] [wsgi:error] [pid 40625:tid 139688744765312]     from psycopg2._psycopg import (                     # noqa
[Thu Nov 09 04:26:45.690032 2023] [wsgi:error] [pid 40623:tid 139688744765312] During handling of the above exception, another exception occurred:
[Thu Nov 09 04:26:45.690033 2023] [wsgi:error] [pid 40622:tid 139688744765312] ModuleNotFoundError: No module named 'psycopg2._psycopg'`

I searched online on how to install psycopg2 and I found these commands `pip install psycopg2-binary` and `sudo pip install psycopg2-binary` but it still does not work. May I know what am I doing wrong?

**Comment by jdetaeye** [comment]
We don't distribute a debian installer for ubuntu 22.   
Using the ubuntu 20 installer on ubuntu 22 won't work.  I think this is what you did, isn't it?

So either you compile from scratch and create your own installer for ubuntu 22.
Either you use a docker container (based on ubuntu 20) to deploy frepple on ubuntu 22.
Either you install on an ubuntu 20 server.

**Reply by seejayyy** [reply]
Hello, thank you so much for your reply. It has helped me a lot. 

I have installed Ubuntu 20 and installed frepple using the written guide on the official website. 

However, when I tried to run the command `sudo frepplectl loaddata demo`, I get the error below. What can I do to resolve this error?

```Traceback (most recent call last):
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedTable: relation "execute_log" does not exist
LINE 1: INSERT INTO "execute_log" ("name", "submitted", "started", "...
                    ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/execute/management/commands/loaddata.py", line 132, in handle
    task.save(using=database)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/base.py", line 739, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/base.py", line 776, in save_base
    updated = self._save_table(
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/base.py", line 881, in _save_table
    results = self._do_insert(cls._base_manager, using, fields, returning_fields, raw)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/base.py", line 919, in _do_insert
    return manager._insert(
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/query.py", line 1270, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1416, in execute_sql
    cursor.execute(sql, params)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: relation "execute_log" does not exist
LINE 1: INSERT INTO "execute_log" ("name", "submitted", "started", "...
                    ^


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/frepplectl", line 102, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/freppledb/execute/management/commands/loaddata.py", line 232, in handle
    task.save(
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/base.py", line 739, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/base.py", line 776, in save_base
    updated = self._save_table(
  File "/usr/share/frepple/venv/lib/python3.8/site-packages/django/db/models/base.py", line 841, in _save_table
    raise ValueError("Cannot force an update in save() with no primary key.")
ValueError: Cannot force an update in save() with no primary key.

**Reply by jdetaeye** [reply]
You have not created the postgres database and/or not created the database tables.

Creating the database can be done with the command "frepplectl createdatabase". Or you use the postgresql command https://www.postgresql.org/docs/current/sql-createdatabase.html

Creating the database tables is done with the command "frepplectl migrate". 

If you would have used a docker container, both the database creation and migration would have been done automatically.

**Comment by seejayyy** [comment]
> You have not created the postgres database and/or not created the database tables.
> 
> Creating the database can be done with the command "frepplectl createdatabase". Or you use the postgresql command https://www.postgresql.org/docs/current/sql-createdatabase.html
> 
> Creating the database tables is done with the command "frepplectl migrate".
> 
> If you would have used a docker container, both the database creation and migration would have been done automatically.

I have tried using the command `sudo frepplectl createdatabase` but it still leads me to the same error.

**Reply by jdetaeye** [reply]
Then the database parameters you configured in your djangosettings.py file are not allowing you to connect to the database.

The installation script https://frepple.com/docs/current/installation-guide/linux-binaries.php#ubuntu-installation-script has steps to create   the user and databases that are in the default djangosettings.py file.

**Comment by seejayyy** [comment]
> Then the database parameters you configured in your djangosettings.py file are not allowing you to connect to the database.
> 
> The installation script https://frepple.com/docs/current/installation-guide/linux-binaries.php#ubuntu-installation-script has steps to create the user and databases that are in the default djangosettings.py file.

I was referring to the installation script while installing frepple and I have even copied them line by line to double check I did not miss out any commands during installation but it still gives the same error.

**Comment by seejayyy** [comment, marked as answer]
I figured it out. I am supposed to do `sudo frepplectl migrate` before `sudo frepplectl loaddata demo`. Now it is working.
